tests/test_coroutine.rs

```diff
@@ -0,0 +1,117 @@
+use futures::channel::oneshot;
+use futures::future::poll_fn;
+use futures::FutureExt;
+use pyo3::coroutine::CoroutineCancel;
+use pyo3::prelude::*;
+use pyo3::py_run;
+use std::task::Poll;
+use std::thread;
+use std::time::Duration;
+
+#[path = "../src/tests/common.rs"]
+mod common;
+
+#[test]
+fn noop_coroutine() {
+    #[pyfunction]
+    async fn noop() -> usize {
+        42
+    }
+    Python::with_gil(|gil| {
+        let noop = wrap_pyfunction!(noop, gil).unwrap();
+        let test = "import asyncio; assert asyncio.run(noop()) == 42";
+        py_run!(gil, noop, test);
+    })
+}
+
+#[test]
+fn sleep_0_like_coroutine() {
+    #[pyfunction]
+    async fn sleep_0() -> usize {
+        let mut waken = false;
+        poll_fn(|cx| {
+            if !waken {
+                cx.waker().wake_by_ref();
+                waken = true;
+                return Poll::Pending;
+            }
+            Poll::Ready(42)
+        })
+        .await
+    }
+    Python::with_gil(|gil| {
+        let sleep_0 = wrap_pyfunction!(sleep_0, gil).unwrap();
+        let test = "import asyncio; assert asyncio.run(sleep_0()) == 42";
+        py_run!(gil, sleep_0, test);
+    })
+}
+
+#[pyfunction]
+async fn sleep(seconds: f64) -> usize {
+    let (tx, rx) = oneshot::channel();
+    thread::spawn(move || {
+        thread::sleep(Duration::from_secs_f64(seconds));
+        tx.send(42).unwrap();
+    });
+    rx.await.unwrap()
+}
+
+#[test]
+fn sleep_coroutine() {
+    Python::with_gil(|gil| {
+        let sleep = wrap_pyfunction!(sleep, gil).unwrap();
+        let test = "import asyncio; assert asyncio.run(sleep(0.1)) == 42";
+        py_run!(gil, sleep, test);
+    })
+}
+
+#[test]
+fn cancelled_coroutine() {
+    Python::with_gil(|gil| {
+        let sleep = wrap_pyfunction!(sleep, gil).unwrap();
+        let test = r#"
+        import asyncio;
+        async def main():
+            task = asyncio.create_task(sleep(1))
+            await asyncio.sleep(0)
+            task.cancel()
+            await task
+        asyncio.run(main())
+        "#;
+        let globals = gil.import("__main__").unwrap().dict();
+        globals.set_item("sleep", sleep).unwrap();
+        let err = gil
+            .run(&pyo3::unindent::unindent(test), Some(globals), None)
+            .unwrap_err();
+        assert_eq!(err.value(gil).get_type().name().unwrap(), "CancelledError");
+    })
+}
+
+#[test]
+fn coroutine_cancel() {
+    #[pyfunction]
+    async fn cancellable_sleep(seconds: f64, mut cancel: CoroutineCancel) -> usize {
+        futures::select! {
+            _ = sleep(seconds).fuse() => 42,
+            _ = cancel.cancelled().fuse() => 0,
+        }
```

Comment:
This works quite nicely, and it seems to me to be a fairly good mapping to Python's `await` keyword potentially raising a `CancelledError`. 👍